src/librustc_mir/transform/uninhabited_enum_branching.rs

```diff
@@ -0,0 +1,129 @@
+//! A pass that eliminates branches on uninhabited enum variants.
+
+use rustc::mir::{
+    BasicBlock, Body, Local, Operand, Place, PlaceBase, Rvalue, StatementKind, TerminatorKind
+};
+use rustc::ty::{Ty, TyCtxt};
+use rustc::ty::layout::{Abi, TyLayout, Variants};
+use crate::transform::{MirPass, MirSource};
+
+pub struct UninhabitedEnumBranching;
+
+fn get_discriminant_local(terminator: &TerminatorKind<'_>) -> Option<Local> {
+    if let TerminatorKind::SwitchInt { discr: Operand::Move(Place { base: PlaceBase::Local(l),
+                                                                    projection: box [] }),
+                                       .. } = terminator {
+        Some(*l)
+    } else {
+        None
+    }
+}
+
+/// Finds all basic blocks which terminate by switching on a discriminant.
+/// Returns a `Vec` of block ids and the `Ty` the discriminant is read from.
+fn find_eligable_blocks<'tcx>(body: &Body<'tcx>) -> Vec<(BasicBlock, Ty<'tcx>)> {
+    let mut blocks_to_update = Vec::new();
+
+    for (bb, block_data) in body.basic_blocks().iter_enumerated() {
+        let terminator = block_data.terminator();
+
+        // only bother checking blocks which terminate by switching on a local
+        if let Some(local) = get_discriminant_local(&terminator.kind) {
+            let stmt_before_term =
+                if block_data.statements.len() > 0 {
```

Comment:
Could use `.then_with` here.

Reply:
Did you mean `.and_then()`?

Reply:
I'm not sure `.and_then()` makes this any more readable because we still need `local` in scope below.

Reply:
https://doc.rust-lang.org/nightly/std/primitive.bool.html#method.then_with :)

Reply:
Oh, gotcha. I was searching the stable docs and only found `Ordering::then_with` which didn't make any sense.